File: params.py
"""
Parameter dataclasses for tessera.
"""

from dataclasses import dataclass, field


@dataclass(frozen=True)
class Argon2Params:
    """
    Parameters for libsodium's Argon2id (crypto_pwhash, alg = ARGON2ID13).

    Defaults are conservative for production use:
    - opslimit=3:           3 passes (iterations / "time cost").
    - memlimit=268435456:   256 MiB of memory, in BYTES.
    - hash_len=64:          64 bytes = 512 bits output.

    Note: unlike the raw Argon2 spec, libsodium's crypto_pwhash FIXES the
    parallelism (lanes) at 1, so there is no parallelism parameter here. This
    is intentional and helps cross-implementation determinism (Python / TS /
    Dart all using libsodium agree byte-for-byte).

    Security invariant: hash_len MUST be ≥ 64 bytes.
    A 64-byte (512-bit) output reduced mod P (256-bit prime) has bias ≤ 2^-256,
    which is computationally negligible. A 32-byte output would have bias ~2^0.
    """

    opslimit: int = 3
    memlimit: int = 256 * 1024 * 1024  # 256 MiB, in bytes
    hash_len: int = 64


@dataclass(frozen=True)
class Params:
    """
    Full parameter set for tessera enrollment and unlock.

    argon2: Argon2id parameters used for answer blinding.
    kdf_context: BLAKE2b KDF domain-separation string. Changing this
                 invalidates all existing vaults.
    max_subsets: Upper bound on C(M, t) subset iterations allowed in unlock.
                 Prevents accidentally creating vaults that take years to unlock.
                 Default 1,000,000 subsets.
    """

    argon2: Argon2Params = field(default_factory=Argon2Params)
    kdf_context: bytes = b"tessera/kek/v1"
    max_subsets: int = 1_000_000

File: ports/dart/lib/src/params.dart
/// Parameter types for tessera (mirrors the Python reference params.py).
library;

import 'dart:convert';
import 'dart:typed_data';

/// Argon2id parameters (libsodium crypto_pwhash; parallelism is fixed at 1).
class Argon2Params {
  /// Passes over memory.
  final int opslimit;

  /// Memory in BYTES.
  final int memlimit;

  /// Output length; security invariant: MUST be >= 64 (bias mod P <= 2^-256).
  final int hashLen;

  const Argon2Params({
    this.opslimit = 3,
    this.memlimit = 256 * 1024 * 1024,
    this.hashLen = 64,
  });
}

/// Full enrollment/unlock parameter set.
class Params {
  final Argon2Params argon2;

  /// BLAKE2b KDF domain separation. Changing it invalidates existing vaults.
  final Uint8List kdfContext;

  /// Upper bound on C(M, t) subset iterations at unlock.
  final int maxSubsets;

  Params({
    this.argon2 = const Argon2Params(),
    Uint8List? kdfContext,
    this.maxSubsets = 1000000,
  }) : kdfContext = kdfContext ?? defaultKdfContext();

  static Uint8List defaultKdfContext() =>
      Uint8List.fromList(utf8.encode('tessera/kek/v1'));
}
